// @progress/kendo-react-data-tools/columnmenu/ColumnMenu.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import { Popup as me } from "@progress/kendo-react-popup";
import { useLocalization as fe } from "@progress/kendo-react-intl";
import { isCompositeFilterDescriptor as de } from "@progress/kendo-data-query";
import { ColumnMenuItem as I } from "./ColumnMenuItem.mjs";
import { ColumnMenuTextFilter as G, ColumnMenuNumericFilter as H, ColumnMenuDateFilter as J, ColumnMenuBooleanFilter as Ce } from "./ColumnMenuFilters.mjs";
import { ColumnMenuFilterLogic as B } from "./ColumnMenuFilterLogic.mjs";
import { ColumnMenuFilterForm as pe } from "./ColumnMenuFilterForm.mjs";
import { columnMenuSortAscending as K, messages as R, columnMenuSortDescending as Q, columnMenuFilterTitle as U } from "../messages/index.mjs";
import { validatePackage as ge, IconWrap as ke } from "@progress/kendo-react-common";
import { packageMetadata as Fe } from "../package-metadata.mjs";
import { sortAscSmallIcon as ve, sortDescSmallIcon as he, filterIcon as V } from "@progress/kendo-svg-icons";
const h = (o) => {
  ge(Fe);
  const f = e.useRef([]), {
    field: t,
    filter: c,
    sort: n = f.current,
    onSortChange: m,
    onFilterChange: r,
    initialFilter: p,
    sortAsc: A = !1,
    sortDesc: T = !1,
    filterContent: y = [],
    expandFilters: b = !1
  } = o, [d, L] = e.useState(!1), [E, q] = e.useState(b), [i, F] = e.useState(t && p ? p(t) : null), S = e.useRef(null), M = e.useRef(null), x = e.useRef(null), D = e.useRef(!1), g = e.useCallback(
    () => (c || []).find((l) => l.filters.every((s) => !de(s) && s.field === t)),
    [c, t]
  ), C = e.useCallback(
    () => {
      L(!1);
    },
    []
  ), X = e.useCallback(
    () => {
      if (D.current) {
        D.current = !1;
        return;
      }
      clearTimeout(x.current), x.current = window.setTimeout(() => {
        C();
      });
    },
    [C]
  ), Y = e.useCallback(
    () => clearTimeout(x.current),
    []
  ), Z = e.useCallback((l) => {
    D.current = d && l.currentTarget === S.current;
  }, [d]), $ = e.useCallback(
    () => {
      L(!d), t && p && (q(d ? !1 : b), F(d ? i : g() || p(t)));
    },
    [g, t, d, i, p, b]
  );
  e.useEffect(
    () => {
      d && M.current && M.current.focus();
    },
    [d]
  );
  const ee = e.useCallback(
    () => q(!E),
    [E]
  ), v = e.useCallback(
    (l, s) => {
      const a = [...n], u = a.findIndex((ue) => ue.field === t), k = { field: t, dir: s };
      u >= 0 ? a[u].dir !== k.dir ? a.splice(u, 1, k) : a.splice(u, 1) : a.push(k), C(), m && m.call(void 0, l, a, t);
    },
    [n, t, m, C]
  ), N = e.useCallback(
    (l, s, a) => {
      const u = { ...i, filters: [...i.filters] }, k = u.filters.indexOf(a);
      u.filters.splice(k, 1, s), F(u);
    },
    [i]
  ), te = e.useCallback(
    (l) => {
      l.preventDefault();
      const s = (c || []).slice(), a = g(), u = a ? s.indexOf(a) : -1;
      u >= 0 ? s.splice(u, 1, i) : s.push(i), C(), r && r.call(void 0, l, s, t);
    },
    [c, i, t, r, C, g]
  ), ne = e.useCallback(
    (l) => {
      const s = (c || []).slice(), a = g(), u = a ? s.indexOf(a) : -1;
      u >= 0 && s.splice(u, 1), p && F(p(t)), C(), r && r.call(void 0, l, s, t);
    },
    [c, t, r, p, C, g]
  ), le = e.useCallback((l) => v(l, "asc"), [v]), oe = e.useCallback((l) => v(l, "desc"), [v]), re = e.useCallback(
    (l) => i && F({ ...i, logic: l.value.logic }),
    [i]
  ), O = n.find((l) => l.field === t) || { dir: "" }, [
    z,
    P,
    W
  ] = y, w = fe(), ie = A || T ? /* @__PURE__ */ e.createElement("div", { className: "k-columnmenu-item-wrapper", key: "sorting" }, A && /* @__PURE__ */ e.createElement(
    I,
    {
      title: w.toLanguageString(K, R[K]),
      iconClass: "k-i-sort-asc-small",
      svgIcon: ve,
      selected: O.dir === "asc",
      onClick: le
    }
  ), T && /* @__PURE__ */ e.createElement(
    I,
    {
      title: w.toLanguageString(Q, R[Q]),
      iconClass: "k-i-sort-desc-small",
      svgIcon: he,
      selected: O.dir === "desc",
      onClick: oe
    }
  )) : null, se = i && y.length > 0 ? /* @__PURE__ */ e.createElement("div", { className: "k-columnmenu-item-wrapper", key: "filtering" }, /* @__PURE__ */ e.createElement(
    I,
    {
      iconClass: "k-i-filter",
      svgIcon: V,
      title: w.toLanguageString(U, R[U]),
      onClick: ee
    }
  ), /* @__PURE__ */ e.createElement(
    pe,
    {
      show: E,
      onSubmit: te,
      onReset: ne
    },
    z && /* @__PURE__ */ e.createElement(z, { filter: i.filters[0], onFilterChange: N }),
    P && /* @__PURE__ */ e.createElement(P, { value: i.logic, onChange: re }),
    W && /* @__PURE__ */ e.createElement(W, { filter: i.filters[1], onFilterChange: N })
  )) : null, _ = [ie, se], ce = /* @__PURE__ */ e.createElement(
    "span",
    {
      ref: S,
      onClick: $,
      onMouseDown: Z,
      className: "k-grid-header-menu k-grid-column-menu" + (g() ? " k-active" : ""),
      key: "button"
    },
    /* @__PURE__ */ e.createElement(ke, { name: "filter", icon: V })
  ), ae = /* @__PURE__ */ e.createElement(
    me,
    {
      anchor: S.current,
      show: d,
      key: "popup",
      popupClass: "k-column-menu k-column-menu-popup"
    },
    /* @__PURE__ */ e.createElement(
      "div",
      {
        ref: M,
        tabIndex: 0,
        onBlur: X,
        onFocus: Y,
        style: { outline: "none" }
      },
      o.itemsRender ? o.itemsRender.call(void 0, _, { closeMenu: C, field: t }) : _
    )
  ), j = [ce, ae];
  return /* @__PURE__ */ e.createElement(e.Fragment, null, o.render ? o.render.call(void 0, j) : j);
}, Te = (o) => {
  const { field: f, ...t } = o, c = e.useCallback(
    (r) => ({ logic: "and", filters: [{ operator: "contains", field: r, value: "" }, { operator: "contains", field: r, value: "" }] }),
    []
  ), n = !!o.field, m = n ? [
    G,
    B,
    G
  ] : [];
  return /* @__PURE__ */ e.createElement(
    h,
    {
      sortAsc: n,
      sortDesc: n,
      filterContent: m,
      initialFilter: n ? c : void 0,
      field: f,
      ...t
    }
  );
}, ye = (o) => {
  const { field: f, ...t } = o, c = e.useCallback(
    (r) => ({ logic: "or", filters: [{ operator: "eq", field: r, value: null }, { operator: "eq", field: r, value: null }] }),
    []
  ), n = !!o.field, m = n ? [
    H,
    B,
    H
  ] : [];
  return /* @__PURE__ */ e.createElement(
    h,
    {
      sortAsc: n,
      sortDesc: n,
      filterContent: m,
      initialFilter: n ? c : void 0,
      field: f,
      ...t
    }
  );
}, Le = (o) => {
  const { field: f, ...t } = o, c = e.useCallback(
    (r) => ({ logic: "or", filters: [{ operator: "eq", field: r, value: null }, { operator: "eq", field: r, value: null }] }),
    []
  ), n = !!o.field, m = n ? [
    J,
    B,
    J
  ] : [];
  return /* @__PURE__ */ e.createElement(
    h,
    {
      sortAsc: n,
      sortDesc: n,
      filterContent: m,
      initialFilter: n ? c : void 0,
      field: f,
      ...t
    }
  );
}, qe = (o) => {
  const { field: f, ...t } = o, c = e.useCallback(
    (r) => ({ logic: "and", filters: [{ operator: "eq", field: r, value: !1 }] }),
    []
  ), n = !!o.field, m = n ? [
    Ce
  ] : [];
  return /* @__PURE__ */ e.createElement(
    h,
    {
      sortAsc: n,
      sortDesc: n,
      filterContent: m,
      initialFilter: n ? c : void 0,
      field: f,
      ...t
    }
  );
};
export {
  qe as ColumnMenuBooleanColumn,
  Le as ColumnMenuDateColumn,
  ye as ColumnMenuNumericColumn,
  Te as ColumnMenuTextColumn
};
