// @progress/kendo-react-pdf/grid/grid-query.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { findElement as t, matchesClasses as r, matchesNodeName as a } from "../common/dom-queries.mjs";
const o = "k-grid-header", i = "k-grid-footer", h = "TABLE", c = a(h), n = (s) => s ? "locked" : "wrap";
class l {
  constructor(e) {
    this.element = e, this.list = t(e, r("k-grid-container"));
  }
  content(e) {
    return t(this.list, r(`k-grid-content${e ? "-locked" : ""}`));
  }
  header(e) {
    return this.headerWrap = this.headerWrap || t(this.element, r(o)), t(this.headerWrap, r(`${o}-${n(e)}`));
  }
  footer(e) {
    return this.footerWrap = this.footerWrap || t(this.element, r(i)), t(this.footerWrap, r(`${i}-${n(e)}`));
  }
  table() {
    return t(this.element, c);
  }
}
export {
  i as FOOTER_CLASS,
  l as GridQuery,
  o as HEADER_CLASS
};
