// @progress/kendo-react-layout/bottomnavigation/BottomNavigationItem.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import t from "prop-types";
import { validatePackage as P, classNames as D, svgIconPropType as T, IconWrap as h } from "@progress/kendo-react-common";
import { packageMetadata as B } from "../package-metadata.mjs";
const g = e.forwardRef((i, f) => {
  P(B);
  const m = e.useRef(null);
  e.useImperativeHandle(
    f,
    () => ({
      element: m.current
    })
  );
  const {
    className: r,
    style: k,
    selected: o,
    disabled: a,
    item: x,
    render: y,
    dataItem: N,
    icon: d,
    svgIcon: b,
    text: p,
    id: C,
    onSelect: s,
    onKeyDown: l,
    index: n,
    tabIndex: E = M.tabIndex
  } = i, R = e.useMemo(
    () => D(
      "k-bottom-nav-item",
      {
        "k-selected": o,
        "k-disabled": a
      },
      r
    ),
    [o, a, r]
  ), w = e.useCallback(
    (c) => {
      s && n !== void 0 && !a && s(c, n);
    },
    [s, n, a]
  ), K = e.useCallback(
    (c) => {
      l && n !== void 0 && !a && l(c, n);
    },
    [l, n, a]
  ), v = y, I = x, u = /* @__PURE__ */ e.createElement(
    "span",
    {
      ref: m,
      className: R,
      style: k,
      role: "link",
      id: C,
      tabIndex: E,
      onClick: w,
      onKeyDown: K,
      "aria-current": o,
      "aria-disabled": a
    },
    I ? /* @__PURE__ */ e.createElement(I, { itemIndex: n, item: N }) : /* @__PURE__ */ e.createElement(e.Fragment, null, (d || b) && /* @__PURE__ */ e.createElement(h, { className: "k-bottom-nav-item-icon", name: d, icon: b, size: "xlarge" }), p && /* @__PURE__ */ e.createElement("span", { className: "k-bottom-nav-item-text", style: { userSelect: "none" } }, p))
  );
  return v !== void 0 ? v.call(void 0, u, i) : u;
}), M = {
  tabIndex: 0
};
g.propTypes = {
  className: t.string,
  style: t.object,
  id: t.string,
  disabled: t.bool,
  selected: t.bool,
  icon: t.string,
  svgIcon: T,
  text: t.string,
  tabIndex: t.number
};
g.displayName = "KendoReactBottomNavigationItem";
export {
  g as BottomNavigationItem
};
