// @progress/kendo-react-inputs/maskedtextbox/utils.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
const r = {
  "#": /[\d\s\+\-]/,
  "&": /[\S]/,
  0: /[\d]/,
  9: /[\d\s]/,
  "?": /[a-zA-Z\s]/,
  A: /[a-zA-Z0-9]/,
  C: /./,
  L: /[a-zA-Z]/,
  a: /[a-zA-Z0-9\s]/
}, n = () => !1, i = (a, e) => a.includeLiterals !== e.includeLiterals || a.mask !== e.mask || a.prompt !== e.prompt || a.promptPlaceholder !== e.promptPlaceholder || !o(a.rules, e.rules), o = (a, e) => {
  if (!!a != !!e)
    return !1;
  if (a === e || !a || !e)
    return !0;
  let t = !0;
  for (const s in a)
    if (a[s] !== e[s]) {
      t = !1;
      break;
    }
  if (t) {
    for (const s in e)
      if (!a.hasOwnProperty(s)) {
        t = !1;
        break;
      }
  }
  return t;
};
export {
  r as defaultRules,
  i as maskingChanged,
  n as returnFalse
};
