// @progress/kendo-react-indicators/badge/Badge.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import o from "prop-types";
import { validatePackage as v, useDir as y, classNames as C, kendoThemeMaps as B } from "@progress/kendo-react-common";
import { packageMetadata as O } from "../package-metadata.mjs";
const f = t.forwardRef((e, g) => {
  v(O);
  const {
    cutoutBorder: m = r.cutoutBorder,
    position: d = r.position,
    align: l = r.align,
    children: b,
    className: c,
    style: k
  } = e, a = t.useRef(null), h = t.useCallback(
    () => {
      a.current && a.current.focus();
    },
    []
  );
  t.useImperativeHandle(g, () => ({
    element: a.current,
    focus: h
  }));
  const M = y(a, e.dir), i = t.useMemo(
    () => e.size === null ? null : e.size || r.size,
    [e.size]
  ), n = t.useMemo(
    () => e.fillMode === null ? null : e.fillMode || r.fillMode,
    [e.fillMode]
  ), s = t.useMemo(
    () => e.themeColor || r.themeColor,
    [e.themeColor]
  ), u = t.useMemo(
    () => e.rounded !== void 0 && e.rounded || "medium",
    [e.rounded]
  ), z = t.useMemo(
    () => C(
      "k-badge",
      {
        "k-badge-sm": i === "small",
        "k-badge-md": i === "medium",
        "k-badge-lg": i === "large",
        [`k-badge-${n}`]: n,
        [`k-badge-${n}-${s}`]: !!(n && s),
        "k-badge-border-cutout": m,
        [`k-badge-${d}`]: d,
        "k-top-start": l.vertical === "top" && l.horizontal === "start",
        "k-top-end": l.vertical === "top" && l.horizontal === "end",
        "k-bottom-start": l.vertical === "bottom" && l.horizontal === "start",
        "k-bottom-end": l.vertical === "bottom" && l.horizontal === "end"
      },
      e.rounded !== null ? [`k-rounded-${B.roundedMap[u] || u}`] : void 0,
      c
    ),
    [
      i,
      n,
      s,
      u,
      m,
      l,
      d,
      c
    ]
  );
  return /* @__PURE__ */ t.createElement("span", { className: z, style: k, dir: M }, b);
});
f.propTypes = {
  className: o.string,
  dir: o.string,
  style: o.object,
  align: o.shape({
    vertical: o.oneOf(["top", "bottom"]),
    horizontal: o.oneOf(["start", "end"])
  }),
  size: o.oneOf([null, "small", "medium", "large"]),
  fillMode: o.oneOf([null, "solid", "outline"]),
  themeColor: o.oneOf([
    null,
    "base",
    "primary",
    "secondary",
    "tertiary",
    "info",
    "success",
    "warning",
    "error",
    "dark",
    "light",
    "inverse",
    "inherit"
  ]),
  rounded: o.oneOf([null, "small", "medium", "large", "full"]),
  position: o.oneOf(["edge", "outside", "inside"]),
  cutoutBorder: o.bool
};
const r = {
  size: "medium",
  fillMode: "solid",
  themeColor: "primary",
  position: "edge",
  align: { vertical: "top", horizontal: "end" },
  cutoutBorder: !1
};
f.displayName = "KendoBadge";
export {
  f as Badge
};
