// @progress/kendo-react-layout/stepper/Step.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2024 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import a from "prop-types";
import { StepperContext as Y } from "./context/StepperContext.mjs";
import { focusFirstFocusableChild as Z, dispatchEvent as M, classNames as ee, IconWrap as T, toIconName as te } from "@progress/kendo-react-common";
import { useLocalization as ae } from "@progress/kendo-react-intl";
import { checkOutlineIcon as se, exclamationCircleIcon as ne } from "@progress/kendo-svg-icons";
import { DEFAULT_ANIMATION_DURATION as oe, NO_ANIMATION as ce } from "./contants.mjs";
import { messages as le, optionalText as ie } from "./messages/index.mjs";
const L = e.forwardRef((O, R) => {
  const {
    // content
    children: V,
    className: g,
    current: d,
    disabled: n,
    focused: E,
    icon: r,
    svgIcon: h,
    index: s,
    isValid: t,
    label: o,
    optional: m,
    style: C,
    tabIndex: me = re.tabIndex,
    text: y,
    ...z
  } = O, {
    animationDuration: u,
    isVertical: k,
    item: D,
    linear: H,
    mode: S,
    numOfSteps: c,
    value: l,
    onChange: b,
    onFocus: f,
    successIcon: _,
    errorIcon: w,
    successSVGIcon: B,
    errorSVGIcon: G
  } = e.useContext(Y), p = e.useRef(null), F = e.useCallback(
    () => {
      p.current && Z(p.current);
    },
    []
  ), v = e.useCallback(
    () => ({
      element: p.current,
      focus: F
    }),
    [F]
  );
  e.useImperativeHandle(R, v);
  const I = !H || s === l - 1 || s === l || s === l + 1, N = S === "labels" || !!r && !!o, P = ae(), U = ((i) => P.toLanguageString(i, le[i]))(ie), W = typeof u == "number" ? u : u !== !1 ? oe : ce, $ = e.useCallback(
    (i) => {
      b && !n && M(
        b,
        i,
        v(),
        { value: s }
      );
    },
    [b, l, n]
  ), j = e.useCallback(
    (i) => {
      f && !n && M(
        f,
        i,
        v(),
        void 0
      );
    },
    [f, n]
  ), K = e.useMemo(
    () => ee(
      "k-step",
      {
        "k-step-first": s === 0,
        "k-step-last": c && s === c - 1,
        "k-step-done": s < l,
        "k-step-current": d,
        "k-step-optional": m,
        "k-step-error": t !== void 0 && !t,
        "k-step-success": t,
        "k-disabled": n,
        "k-focus": E
      },
      g
    ),
    [s, c, l, d, m, n, E, t, g]
  ), q = e.useMemo(
    () => ({
      maxWidth: k ? void 0 : `calc(100% / ${c})`,
      maxHeight: k ? `calc(100% / ${c})` : void 0,
      pointerEvents: I ? void 0 : "none",
      ...C
    }),
    [k, c, C, I]
  ), A = t ? _ : w, x = A ? /* @__PURE__ */ e.createElement(
    "span",
    {
      className: "k-step-indicator-icon " + A,
      "aria-hidden": "true"
    }
  ) : /* @__PURE__ */ e.createElement(
    T,
    {
      className: "k-step-indicator-icon",
      name: t ? "check-circle" : "exclamation-circle",
      icon: t ? B || se : G || ne
    }
  ), J = /* @__PURE__ */ e.createElement(e.Fragment, null, S !== "labels" ? /* @__PURE__ */ e.createElement(
    "span",
    {
      className: "k-step-indicator",
      "aria-hidden": !0,
      style: { transitionDuration: W + "ms" }
    },
    r || h ? !N && t !== void 0 ? x : /* @__PURE__ */ e.createElement(T, { className: "k-step-indicator-icon", name: r && te(r), icon: h }) : t !== void 0 ? x : /* @__PURE__ */ e.createElement("span", { className: "k-step-indicator-text" }, y || s + 1)
  ) : null), Q = (o || N || m) && /* @__PURE__ */ e.createElement("span", { className: "k-step-label" }, o && /* @__PURE__ */ e.createElement("span", { className: "k-step-text" }, o), N && t !== void 0 && x, m && /* @__PURE__ */ e.createElement("span", { className: "k-step-label-optional" }, U)), X = /* @__PURE__ */ e.createElement(e.Fragment, null, J, Q);
  return /* @__PURE__ */ e.createElement(
    "li",
    {
      ref: p,
      className: K,
      style: q,
      ...z
    },
    /* @__PURE__ */ e.createElement(
      "a",
      {
        className: "k-step-link",
        title: o || void 0,
        onClick: $,
        onFocus: j,
        "aria-current": d ? "step" : void 0,
        "aria-disabled": n || !I || void 0,
        "aria-invalid": t !== void 0 && !t || void 0
      },
      D ? V : X
    )
  );
});
L.propTypes = {
  children: a.any,
  className: a.string,
  // content: PropTypes.any,
  current: a.bool,
  disabled: a.bool,
  icon: a.string,
  index: a.number,
  isValid: a.bool,
  label: a.string,
  optional: a.bool,
  style: a.object,
  tabIndex: a.number,
  text: a.string
};
const re = {
  tabIndex: 0
};
L.displayName = "KendoStep";
export {
  L as Step
};
